common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+@ThreadSafe
+@Slf4j
+/**
+ * This is a continuation token based async iterator implementation. This class takes a function that when completed will 
+ * have next batch of results with continuation token. 
+ * This class determines when to call the next iteration of function (if all existing results have been exhausted) and 
+ * ensures there is only one outstanding call. 
+ */
+public class ContinuationTokenAsyncIterator<T, U> implements AsyncIterator<U> {
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final Queue<U> queue;
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final AtomicReference<T> token;
+    private final Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function;
+    @GuardedBy("lock")
+    private CompletableFuture<Void> outstanding;
+    private final AtomicBoolean canContinue;
+    private final AtomicBoolean isOutstanding;
+    
+    public ContinuationTokenAsyncIterator(Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function, T tokenIdentity) {
+        this.function = function;
+        this.token = new AtomicReference<>(tokenIdentity);
+        this.queue = new LinkedBlockingQueue<>();
+        this.outstanding = CompletableFuture.completedFuture(null);
+        this.canContinue = new AtomicBoolean(true);
+        this.isOutstanding = new AtomicBoolean(false);
+    }
+
+    @Override
+    public CompletableFuture<U> getNext() {
+        final T continuationToken;
+        boolean toCall = false;
+        synchronized (lock) {
+            // if the result is available, return it without making function call
+            if (!queue.isEmpty()) {
```

Comment:
So you check this queue here and if it's not empty, you return the first element, else you invoke the function and add to the queue and reinvoke this method.

I think you can greatly simplify this class by removing the queue altogether and invoking your function (in a loop, maybe) and returning the result. You can also get rid of a few other fields as well.

Reply:
Actually the whole idea is that each  function call can give you a bunch of results. 
We only want income function again once we exhaust all available results. 
Hence the whole queue logic.
If it was retrieving one element at a time, we would have called getNext in a loop. 

Also we have  supported concurrent getNext calls on the same async iterator. Which would hard to achieve with a future.loop call

Reply:
Ok, I missed the `addAll`. I thought it was just `add`.